### PR TITLE
Replace PCRE 1 with PCRE 2

### LINKED ISSUE #392: pcre library is obsolete, please consider to switch to pcre2
The PCRE library is now abandoned and not maintained anymore. That library should not be used, especially when the input can come from an untrusted source.

There is a newer library implemented (pcre2), which should be used instead (documentation is here: https://pcre.org/current/doc/html/

And also Mageia Linux planned to drop old pcre library like some others distributions.

Regards,
David

### PR DESCRIPTION
**Pull request by loqs**
Closes https://github.com/dlundquist/sniproxy/issues/392.

### CODE REVIEW

File: src/backend.c [hunk 1]
```diff
@@ -128,11 +128,16 @@ lookup_backend(const struct Backend_head *head, const char *name, size_t name_le
 
     STAILQ_FOREACH(iter, head, entries) {
         assert(iter->pattern_re != NULL);
-        if (pcre_exec(iter->pattern_re, NULL,
-                    name, name_len, 0, 0, NULL, 0) >= 0)
+        md = pcre2_match_data_create_from_pattern(iter->pattern_re,NULL);
```
Comment (dlundquist) [inline src/backend.c:131]: When there are multiple entries in the list this will leak memory.

File: src/backend.c [hunk 2]
```diff
@@ -96,13 +95,13 @@ add_backend(struct Backend_head *backends, struct Backend *backend) {
...
 init_backend(struct Backend *backend) {
     if (backend->pattern_re == NULL) {
-        const char *reerr;
-        int reerroffset;
+        int reerr;
+        PCRE2_SIZE reerroffset;
 
         backend->pattern_re =
-            pcre_compile(backend->pattern, 0, &reerr, &reerroffset, NULL);
+            pcre2_compile(backend->pattern, PCRE2_ZERO_TERMINATED, 0, &reerr, &reerroffset, NULL);
```
Comment (dlundquist) [inline src/backend.c:102]: clang complains:
```
backend.c:102:34: warning: pointer targets in passing argument 1 of ‘pcre2_compile_8’ differ in signedness [-Wpointer-sign]
  102 |             pcre2_compile(backend->pattern, PCRE2_ZERO_TERMINATED, 0, &reerr, &reerroffset, NULL);
      |                           ~~~~~~~^~~~~~~~~
      |                                  |
      |                                  char *
```

File: src/backend.c [hunk 3]
```diff
@@ -128,11 +128,16 @@ lookup_backend(const struct Backend_head *head, const char *name, size_t name_le
 
     STAILQ_FOREACH(iter, head, entries) {
         assert(iter->pattern_re != NULL);
-        if (pcre_exec(iter->pattern_re, NULL,
-                    name, name_len, 0, 0, NULL, 0) >= 0)
+        md = pcre2_match_data_create_from_pattern(iter->pattern_re,NULL);
+        if (pcre2_match(iter->pattern_re,
+                    name, name_len, 0, 0, md, 0) >= 0)
```
Comment (dlundquist) [inline src/backend.c:133]: Clang complains:
```
backend.c:133:21: warning: pointer targets in passing argument 2 of ‘pcre2_match_8’ differ in signedness [-Wpointer-sign]
  133 |                     name, name_len, 0, 0, md, 0) >= 0)
      |                     ^~~~
      |                     |
      |                     const char *
```

**Review by dlundquist — DISMISSED** [review]
Thank you, I took a stab at this keeping libpcre3 support so I didn't break existing users  [pcre2](https://github.com/dlundquist/sniproxy/tree/pcre2). Let me know if you want to fix this up and keep libpcre (legacy) support or I should just merge mine.

File: src/backend.c (hunk 1 above)
Comment (loqs) [reply src/backend.c]: I think this is now resolved by moving the call to init_backend and the free to remove_backend.

**Comment by loqs** [comment]
> Thank you, I took a stab at this keeping libpcre3 support so I didn't break existing users [pcre2](https://github.com/dlundquist/sniproxy/tree/pcre2). Let me know if you want to fix this up and keep libpcre (legacy) support or I should just merge mine.

I think I have resolved the issues with this apart from the fact it does not maintain libpcre3 support.  If you want to merge yours I think that would be the best solution as you are more familiar with the code base and have to maintain it. Or I can add libpre3 support back to this.

**Comment by loqs** [comment]
Resolved by https://github.com/dlundquist/sniproxy/commit/7fdd86c054a21f7ac62343010de20f28645b14d2

**Pull request closed without merging**